Source/Core/Core/HW/EXI/BBA/XLINK_KAI_BBA.cpp

```diff
@@ -0,0 +1,298 @@
+  {
+    ERROR_LOG(SP1, "Activate(): failed to send connect message to XLink Kai client");
+  }
+
+  INFO_LOG(SP1, "BBA initialized.");
+
+  return RecvInit();
+}
+
+void CEXIETHERNET::XLinkNetworkInterface::Deactivate()
+{
+  // Send d; to tell XLink we want to disconnect cleanly
+  // disconnect;optional_locally_unique_name;optional_padding
+  std::string cmd =
+      "disconnect;" + m_client_identifier + ";0000000000000000000000000000000000000000000";
+  const auto size = u32(cmd.length());
+  u8 buffer[255] = {};
+  memmove(buffer, cmd.c_str(), size);
+
+  DEBUG_LOG(SP1, "SendCommandPayload %x\n%s", size, ArrayToString(buffer, size, 0x10).c_str());
+
+  if (m_sf_socket.send(buffer, size, m_sf_recipient_ip, m_dest_port) != sf::Socket::Done)
+  {
+    ERROR_LOG(SP1, "Deactivate(): failed to send disconnect message to XLink Kai client");
+  }
+
+  NOTICE_LOG(SP1, "XLink Kai BBA deactivated");
+
+  // Disable socket blocking
+  m_sf_socket.setBlocking(false);
+
+  // Set flags for clean shutdown of the read thread
+  m_read_enabled.Clear();
+  m_read_thread_shutdown.Set();
+
+  // Detach the thread and let it die on its own (leaks ~9kb of memory)
+  m_read_thread.detach();
+
+  // Close the socket
+  m_sf_socket.unbind();
+}
+
+bool CEXIETHERNET::XLinkNetworkInterface::IsActivated()
+{
+  return m_sf_socket.getLocalPort() > 0;
+}
+
+bool CEXIETHERNET::XLinkNetworkInterface::SendFrame(const u8* frame, u32 size)
+{
+  // Is the BBA actually connected?
+  if (!m_bba_link_up)
+  {
+    // Has the user been told the connection failed?
+    if (!m_bba_failure_notified)
+    {
+      OSD::AddMessage("XLink Kai BBA not connected", 30000);
+      m_bba_failure_notified = true;
+    }
+
+    // Return true because this isnt an error dolphin needs to handle
+    return true;
+  }
+
+  // Overwrite buffer and add header
+  memmove(m_out_frame, "e;e;", 4);
+  memmove(m_out_frame + 4, frame, size);
+  size += 4;
+
+  // Only uncomment for debugging, the performance hit is too big otherwise
+  // INFO_LOG(SP1, "SendFrame %x\n%s", size, ArrayToString(m_out_frame, size, 0x10).c_str());
+
+  if (m_sf_socket.send(m_out_frame, size, m_sf_recipient_ip, m_dest_port) != sf::Socket::Done)
+  {
+    ERROR_LOG(SP1, "SendFrame(): expected to write %u bytes, but failed, errno %d", size, errno);
+    return false;
+  }
+  else
+  {
+    m_eth_ref->SendComplete();
+    return true;
+  }
+}
+
+void CEXIETHERNET::XLinkNetworkInterface::ReadThreadHandler(
+    CEXIETHERNET::XLinkNetworkInterface* self)
+{
+  sf::IpAddress sender;
+  u16 port;
+
+  while (!self->m_read_thread_shutdown.IsSet())
+  {
+    if (!self->IsActivated())
+      break;
+
+    // XLink supports jumboframes but Gamecube BBA does not. We need to support jumbo frames
+    // *here* because XLink *could* send one
+    std::size_t bytes_read = 0;
+    if (self->m_sf_socket.receive(self->m_in_frame, std::size(self->m_in_frame), bytes_read, sender,
```

Comment:
How (un)likely is it that this `receive` above returns garbage, or rather: less than 4 bytes read? Depending on the answer, it might be worthwhile/necessary to add more checks before we attempt to copy data into `temp_check` below.

Also, this prints an error to the log, but continues anyways. Should this `break`/ `continue` (or even `return`) in case reading fails?
(Similarly, if this needs to be changed in any way, the `send` below also does that)